### ISSUE TITLE
"Go to Declaration" does not work when the type was asserted to a different type

### ISSUE THREAD

**Issue by sonnguyen9800** [JohnnyMorganz/luau-lsp#782]
### Problem
When using the Luau Language Server (johnymorgan) with VSCode, I cannot use the "Go to Declaration" functionality for functions defined on tables with metatables and type annotations. This occurs in cases where functions are defined for a table using a type that employs setmetatable for object-like behavior.

### Reproduction

Here’s an example:
```
type AccountImpl = {
    __index: AccountImpl,
    new: (name: string, balance: number) -> AccountType,
    deposit: (self: AccountType, credit: number) -> (),
    withdraw: (self: AccountType, debit: number) -> (),
}

type AccountType = typeof(setmetatable({} :: { name: string, balance: number }, {} :: AccountImpl))

local MyAccount: AccountImpl = {} :: AccountImpl
MyAccount.__index = MyAccount

function MyAccount.new(name: string, balance: number): AccountType
    local self = {}
    self.name = name
    self.balance = balance
    return setmetatable(self, MyAccount)
end

function MyAccount:deposit(credit: number)
    self.balance += credit
end

function MyAccount:withdraw(debit: number)
    self.balance -= debit
end

local account = MyAccount.new("Alexander", 500)
account:withdraw(100) -- cannot go to declaration
```

### Expected Behavior
When right-clicking ```account:withdraw(100)``` or ```MyAccount.new(...)``` and selecting "Go to Declaration," the LSP should navigate to the respective function definitions (```MyAccount:withdraw``` or ```MyAccount.new```).

### Observed Behavior
"Go to Declaration" does not work for either the method (withdraw) or the function (new).

### Additional Information

Luau Language Server version: 1.32.4
VSCode version: 1.93.1
Reproduction environment: Window

P/S
- I have tried adding explicit type annotations, but the issue persists.
- If I use **Nightrains's** Roblox-LSP then the issue no longer persist

**Update**
- I also cannot find "references" & "go to definitions".

**Update #2**

So basically I cannot find "references" or "go to definition" if tables assigned to any custom type

The code below is working

```
local Sample = {}
Sample.__index = Sample

function Sample.HelloWorld(): () 
    print("LOL")
end

Sample.HelloWorld() -- Can go to the definition
return Sample
```
While the other is not

```
local Sample = {} :: ACustomType
Sample.__index = Sample

function Sample.HelloWorld(): () 
    print("LOL")
end

Sample.HelloWorld() -- Can't go to the definition
return Sample
```

**Comment by JohnnyMorganz** [comment]
We rely on Luau's type tracking for function definitions and go to declaration.

When you apply a type assertion, then the Luau type checker doesn't ever care / track your definition of `Sample.HelloWorld` anymore when you call the function. It is going to try and find the definition of `HelloWorld` inside of `ACustomType`.

I don't think there is much we can do here to solve this, because really it is working as intended

**Comment by JohnnyMorganz** [comment]
I've been thinking about this, but I don't think there's anything we can do to fix this here. The type assertion means the type checker has no idea of the real definitions anymore.

Going to close this as working as intended. 

If you're looking for a working OOP example, the Luau docs were recently updated with a new structure: https://github.com/luau-lang/site/pull/49/files
